### PR TITLE
Simplify generated code module structure

### PR DESCRIPTION
**Pull request by swallez**
Remove the `generated` directory and the various re-exports.

- namespace docs have been extracted from the respective `mod.rs` to
  `.md` files in api_generator/docs, and are inserted in the generated
  source file.
- namespace sources are generated directly in the `src` directory
- the generator can now insert generated sections in regular source
  files based on `GENERATE-BEGIN` and `GENERATE-END` markers. This is
  used to generate the namespace module list in the main `lib.rs` and
  merge generated with manual code in `params.rs`
- generated and merged source files are listed in `src/.generated.toml`
  so that the generator can cleanup on start.

Before considering merging generated code with special markers we
experimented with the `include!()` macro, but this didn't work well:
- rustfmt will not reformat include files
- `mod` statement look for files relative to the source file, so
  include files have to be located in the same directory as their
  includer, while locating them in a dedicated directory would have
  provided a nicer organization.

Merging generated sections provides a more natural source code layout.

### CODE REVIEW

File: api_generator/src/bin/run.rs [hunk 1]
```diff
@@ -28,9 +28,9 @@ use std::{
 };
 
 fn main() -> Result<(), failure::Error> {
-    // This must be run from the src root directory, with cargo run -p api_generator
+    // This must be run from the repo root directory, with cargo run -p api_generator
```
Comment (russcam) [inline api_generator/src/bin/run.rs:31]: Guess we can update this now to `cargo make generate-api` 😃

File: api_generator/src/generator/mod.rs [hunk 2]
```diff
@@ -409,94 +422,84 @@ pub fn generate(
...
-// cargo run -p api_generator
-// -----------------------------------------------
-"
-        .as_bytes(),
+    fs::write(
+        generated,
+        toml::to_string_pretty(&GeneratedFiles {
+            written: written_files,
+            merged: merged_files,
+        })?,
```
Comment (russcam) [inline api_generator/src/generator/mod.rs:500]: should we make `written_files` and `merged_files` be ordinal ordered, for consistent ordering? I understand they don't need to be, but I ran `cargo make `generate-api` on my machine and the order of entries in `.generated.toml` is different (possible difference in default string ordering implementation on Windows vs. non-Windows)

**Review by russcam — CHANGES_REQUESTED** [review]
The change looks good overall. I left some comments about modules that are now both public and re-exported at the crate root

File: elasticsearch/src/lib.rs [hunk 3]
```diff
@@ -353,24 +353,31 @@ type _DoctestReadme = ();
 #[macro_use]
 extern crate dyn_clone;
 
-pub mod async_search;
 pub mod auth;
+pub mod cert;
+pub mod client;
+pub mod error;
```
[the remark this answers is not in the document]
Comment (swallez) [reply elasticsearch/src/lib.rs]: Good catch, I overlooked this while reorganizing `mod` declarations. Fixed in #9ba6ca2

File: api_generator/src/generator/mod.rs (hunk 2 above)
Comment (swallez) [reply api_generator/src/generator/mod.rs]: Good point, replaced by a `BTreeSet` in 9ba6ca2 to have a stable ordering.

File: api_generator/src/bin/run.rs (hunk 1 above)
Comment (swallez) [reply api_generator/src/bin/run.rs]: Indeed 😉 - done in 00efb77

**Pull request merged**